### ISSUE TITLE
Develop Release Feautures List and Release Date

### ISSUE THREAD

**Issue by thadius856** [RaspberryPints/RaspberryPints#33]
What features/tasks do we need to complete before initial release? Brainstorm here.

After the list is built, we can develop a timeframe.

**Comment by thadius856** [comment]
****To-Do List****

**Initial Release:**
- [ ] Completed RPi installation and configuration instructions, step-by-step, with images/screenshots.
- [ ] FAQ entries pre-written.
- [ ] FAQ entries proofread.
- [x] Ability to hide columns.
- [ ] Admin panel functionality to add beer recipes, put recipes on tap, etc.
- [x] Ability to easily personalize brewery logo and center header.
- [x] Admin panel placeholders for: flow meters, statistics, additional personalization options
- [x] Feedback submission feature in admin panel.
- [x] Bug reporting feature in admin panel.
- [ ] Tap list tested at minimum on: 1080x1920, 1200x1920, and 1920x1080.
- [x] New free/open source background image.
- [ ] All required future SQL database tables built and designed to be extensible.
- [ ] Add error handling for zero kegs set to active ("You don't have any beers added yet. -br- Perhaps you should add some?").
- [ ] Add "Full!" and "Gone!" keg images.
- [ ] Add error handling for failure to connect to the database ("Database not found").
- [ ] Replace SRM and IBU images (last remaining Kegerface asset).
- [ ] Make SRM images auto-tint based on an SRM-to-RGB lookup table (precision down to 0.1 SRM, 10cm path, 40 SRM max).
- [ ] Re-license under an appropriate free license with attribution and non-commercial use.
- [x] Move image heights from HTML to CSS (for 4k support)

**Second Release:**
- [ ] Full flow meter pulse counting support.
- [ ] Flow meter calibration scheme in admin panel.
- [ ] Keg level polling and live keg level updates.
- [ ] Updated instructions to include breakout board assembly and flow meter wiring.
- [ ] Support for RFID reader and drinker accounts.

**Future Releases:**
- [ ] Support for metric units.
- [ ] Support for different keg sizes.
- [ ] Support for temperature probes.
- [ ] Support for motion sensors (to blank screen when not in use).
- [ ] Support for "Coming Soon", "Fermenting", "Force Carbing", "Carbing Up", or similar section (separate table).
- [ ] Support for "On Deck", "Next Brew Day", "Brewing Soon", or similar section (separate table).
- [ ] Support for "To-do", "Brewing Items", "Brewing Actions", or similar section (separate table).
- [ ] Support for multiple layouts of taplist (horizontal orientation, two column horizontal orientation, etc).
- [ ] Support for untapped QR codes.

**Comment by thadius856** [comment]
Reserved for Timeline.

**Comment by Boombrewer** [comment]
What was the consensus about adding the flow meters? Was that first release or second? I personally would like to see it ASAP but also think second release is more practical and serves more individuals. 

So what about this?

First release:
- display only with admin function similar to what's available now
- minimal personalization (just logo and other small stuff)

Second release:
- support for flow meters
- associated analytical options
- improve on personalization options based on feedback from first release

Third release:
- support for temperature probes and associated analytical options

I know there's talk of isolation valves for "safety" (i.e. Roaming cats) or portioning but I'd guess that would be a significant amount of worked based solely on the amount of options available for user interface. 

Just my .02 since I do no coding.   : )
Please keep up the awesome work for us lowly knuckle draggers. You may not be able to or have any desire to sell it, but I'd organize a way for those who reap the benefits to reimburse those who carry the burden. (I'll start that thread after first release when hundreds or thousands are singing your praises!)

**Comment by Skemp45** [comment]
Bare minimum for release, 

Main Site-

-fully functional front end with NO bugs that we can find. (there will always be thing our users find over time)

-Ability to disable the keg column- I don't think we should have a functional keg and temp system for the first release. That should be something our users anticipate and this will cause for user interaction and get the buzz going about the project. 

Admin Panel-

-Ability to Add, Edit and remove beers. 

-ability to change brewery logo and beers on tap text.

-I want to leave links for whats to come again to cause anticipation and for users to give feedback. 

-The best thing about the android market is user participation and I REALLY want users to be involved with this, that will help the product grow and make this become everything we want it to be.

**Comment by agenthubcap** [comment]
I need to pick up some resistors today to wire in the flow meter for testing, but I'm 80% done with the flow meter service.  

If it makes it in to the first release, great.  Have we been designing the view to support an asynchronous update from an external service?  If not, then it should probably wait for the second release.

Sent from Yahoo Mail on Android

**Comment by thadius856** [comment]
> So what about this?
> 
> First release:

Agree with these.

> Second release:
> - support for flow meters
> - associated analytical options
> - improve on personalization options based on feedback from first release

Placeholders would be nice in the first release.

> Bare minimum for release,

Agree with all of these.

> If it makes it in to the first release, great.  Have we been designing the view to support an asynchronous update from an external service?  If not, then it should probably wait for the second release.

In English?

**Comment by thadius856** [comment]
> I need to pick up some resistors today to wire in the flow meter for testing, but I'm 80% done with the flow meter service. 

I'd recommend checking with the guys on the official Raspberry Pi forum, just in case I calculated the wrong ratings on those resistors.

**Comment by thadius856** [comment]
First draft of the feature list coming. Comments? Changes?

**Comment by agenthubcap** [comment]
The view needs to poll for updates or else it won't ever show the levels dropping.  I'd recommend an asynchronous call (something like an Ajax call would work).  

Basically, the part of the page that renders the level is built up in code vs html.

Sent from Yahoo Mail on Android

**Comment by Skemp45** [comment]
Yes, the time and the Keg levels will need to be polled for updates or refreshed.

**Comment by Skemp45** [comment]
nvm I remember we removed time and replaced with the raspberry pints logo

**Comment by Skemp45** [comment]
Feedback submission feature in admin panel.
Bug reporting feature in admin panel.

done and done

It will send directly to support@raspberrypints which cc's to all of us

**Comment by JayUnt** [comment]
Should we put this list into a wiki page instead of an issue? or would that prevent talking back and forth for discussion?

I finished the first checkbox "Ability to hide columns." for the database and the front end.  We just need to add it to the admin.

**Comment by thadius856** [comment]
Wiki would indeed prevent discussion. And email alerts with easy replies. ;)

Will go through the HBT thread tonight and see if I missed any feature requests.

**Comment by JayUnt** [comment]
very true.  Issue it is!

On Mon, Jan 13, 2014 at 8:24 PM, thadius856 notifications@github.comwrote:

> Wiki would indeed prevent discussion. And email alerts with easy replies.
> ;)
> 
> Will go through the HBT thread tonight and see if I missed any feature
> requests.
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/RaspberryPints/RaspberryPints/issues/33#issuecomment-32230733
> .

## 

Jason Unterman

**Comment by thadius856** [comment]
> The view needs to poll for updates or else it won't ever show the levels dropping.  I'd recommend an asynchronous call (something like an Ajax call would work). 
> 
> Basically, the part of the page that renders the level is built up in code vs html.

Huh. Ok, I'll add it to the list. I was going to be ghetto and just use a Chromium addon to auto-refresh every x seconds, but since you know what you're doing, we'll run with the idea.

> Feedback submission feature in admin panel.
> Bug reporting feature in admin panel.

Will test momentarily.

**Comment by JayUnt** [comment]
polling i simple.  Once we get the amount to update automatically and we timestamp that amount, we can poll at any interval.  I was thinking we could poll every 15 - 30 mins and if there was any activity, we could ramp it up to 2 - 5 minutes for a certain amount of time.  most of the time no updates will be made, but if there are, we can assume there will be other close by.  We could even make those numbers configurable.

**Comment by Skemp45** [comment]
I have to add jay and your email to the cc it's only mine right now as I was testing it

**Comment by thadius856** [comment]
Combed through the entire HBT thread and added every feature I saw requested or talked about (within reason). Categorized them as initial, second or future releases.

Couple easy items added to the initial release, ones that we probably already assumed we'd be doing. I added some low-hanging fruit for myself to tick off, since you guys are zooming past me.

Speaking of which, I updated the background (and pushed the commit), then ticked it off. Now the only assets left from Kegerface are the SRM images and the hops.

SRM image re-work added to the list. Jay, if I provide a SRM overlay with no outline and only white fill in the beer portion, can we auto-tint the color based on a look-up table? If so, can you help me with formatting the formatting of that lookup table?

**Comment by Skemp45** [comment]
I added a file in admin called update_personalize.php this pushes the updates to the db for hiding columns. I also made the necessary changes to personalize.php to support column hiding. I just need to figure out how to make it all submit in 1 form instead of 5 different ones. I could not stop it from updating every rows config value when you changed just one.

**Comment by Skemp45** [comment]
Added update_header_text.php which pushes update to table profile that contains logo location and header info. also updated index.php and  personalize.php  to support changes

**Comment by Skemp45** [comment]
Repost but posting here too,  

I would suggest deleting completely and re downloading all files and uploading. I made a ton of changes last night and this morning. I then tested by deleting all my files on my server re downloading them and only making updated to includes/config.php, admin/includes/config.php, admin/includes/conn.php  and imported both sql files and BOOM site loads and works, admin login works everything displays properly. and you can now change header text, brewery logo which updates in both the admin and home page and you can switch columns on and off.

**Comment by Skemp45** [comment]
I just added supporting code so users can change their background image. The only problem is it will remove the original image so I need to write a button next to it that says restore default.

**Comment by JayUnt** [comment]
I know you can change the "tint" of an image, but that has to do with putting a colored, opaque div on top of it. I'm not sure it will work if the image is not square.  

We might be able to use that css masking i talked about a while back.  Make a mask for the shape of the cup, and then the background color of that div is the color you want,  so only the shape of the cup comes through.  Ive never made a mask before but it doesnt look to bad https://www.webkit.org/blog/181/css-masks/.  Ill take a look at it tonight.

**Comment by thadius856** [comment]
> Ill take a look at it tonight.

Thanks. Between having less time and less skill, I'm left in y'alls dusty on this project right now.

I'll dig in this weekend and fuck some shit up, but until then, I'm more or less worthless. Think I could make a decent Project Manager, though. ;)

**Comment by JayUnt** [comment]
You're doing a great job. I'm going to be useless when it comes to setting
to and reading from the flow meter. That's all you.

**Comment by thadius856** [comment]
Unfortunately, there's no "setting to" the flow meter. You provide it +5V and gnd, it sends back pulses at -5V. Zero customization --- all calibration must be done software-side.

I'll probably be just re-using code that's already out there.

As long as I can get the end result of the pour to the `pour` table, that's all that matters, right?

**Comment by JayUnt** [comment]
Yeah. If it's in the pour take I can use it.

But I still don't get what it does. Is it a certain amount of pulses for
different amounts? Or is it the length of the pulse you are looking at?

The most I've done with electronics/writing is setting up a stc-1000. Even
that took me a while until I realized what I was supposed to do. I've just
never done that sorry if thing. I can't wait for you guys to write a how-to
for it and I'll get to learn more about it.

**Comment by thadius856** [comment]
Negative. Its a finned rotor in-line with the fluid. Every 1/x number of
turns, the contactors line up and the +5v line is shorted to the GPIO wire.

We poll the GPIO pin at a specified interval, count the GPIO pulses, and
know how many times the rotor spun.

We can either assume 867000 pulses per litre as the spec states, or we can
pour a liter (or two) as a control and  calculate our real-world,
calibrated pulse rate.

Going back to the pour, we count the pulses,  apply the correction factor,
and record.

Seems straight forward.

**Comment by thadius856** [comment]
*867000 pulses per litre spec

**Comment by thadius856** [comment]
**6000 you stupid phone!

**Comment by JayUnt** [comment]
Ahh nice. That actually does make sense. Very cool.

**Comment by Skemp45** [comment]
So.... Good news, I tested resolution and how it loads on an extreme scale... Looked perfect!! 
![image](https://f.cloud.github.com/assets/6318869/1917475/5a99a8a8-7d99-11e3-9358-b59d71d048ae.jpg)

**Comment by Skemp45** [comment]
![image](https://f.cloud.github.com/assets/6318869/1917487/bfd4e0c0-7d99-11e3-927c-0e94d02e2751.jpg)

**Comment by thadius856** [comment]
I can see in your image that the SRM image has problems with the white border being pixelized.

That's one of the last Kegerface assets. It'll be replaced before first release. Hopefully I'll find time to sit down this weekend and crank out a new one.

Looks cool on that scale, though that's not more than 1920x1080 unless you have a super-expensive 1600x1200 business projector.

My 2560x1600 monitors arrive this evening. I'm curious to see how this looks in both orientations. :D

**Comment by thadius856** [comment]
Added FAQ items to list above. I'd like to write those.

**Comment by Skemp45** [comment]
Also shouldn't all the beers fit on the screen? We should use an identifier of some type to recognize screen size and auto just to that screen size

**Comment by Skemp45** [comment]
Auto adjust*

**Comment by thadius856** [comment]
That would be nice, but I wouldnt say critical. Users can always add or remove beers, or rotate the screen.

Would be a LOT of coding to change the auto-fill images to resize (some have x, y pixel offsets).

**Comment by JayUnt** [comment]
Im not sure we can do the srm icon using a single image and a color.  the problem is we want to show parts of one image and parts of the background.  From what i read masks use to mask images but we are trying to mask a background color.

Another solution i am looking into is using an actual font icon for the srm.  I found this free font icon pack that has a stein as a character.  If we use this, we would be able to make it any size using font-size and color using... you guessed it color.  If you go to this page (http://fontawesome.io/icons/) and search the page for beer, you will see the icon.  Im going to test it out on my local and then upload a screenshot for you guys to see.  if you like it, ill commit.

**Comment by Skemp45** [comment]
http://fontawesome.io/icon/beer

this it? 
Thats genius if we can get it to work!!!

**Comment by JayUnt** [comment]
That link didnt work for me but this one does.  http://fontawesome.io/icon/beer/

So i have two versions setup.  One is using it normally:

![srm icon](https://f.cloud.github.com/assets/35524/1927034/05965dc6-7e50-11e3-85d1-ae599aa9b932.png)

And this next one i was able to mock a white border by putting a slightly larger white one right behind it,

![srm icon with border](https://f.cloud.github.com/assets/35524/1927047/33159aa0-7e50-11e3-8d31-5883316e808f.png)

I personally like the one with the border better.  Im going to check on the licenseing on this to make sure its good for us to use, but from what ive seen its part of the twitter bootstrap which is think is friendly.

We will then need a way to associate colors with srm numbers, and even allow the user to select a different color if they wanted red for fruit juice or white for seltzer or anything like that,

**Comment by JayUnt** [comment]
looks like we are golden with the license.  "Font Awesome is fully open source and is GPL compatible. You can use it for commercial projects, open source projects, or really just about whatever you want."  http://fontawesome.io/license/

**Comment by Boombrewer** [comment]
I have noticed some screen sizing issues. But even I could figure out how to adjust the zoom to make it work. Interesting enough, I've noticed on occasion the page won't let me scroll. I notice it more on my laptop than my iPad and I haven't gone through all the browsers to see if it's persistent.

**Comment by Skemp45** [comment]
Yeah, might be a fix issue. If you have a mouse click do a middle click on the scroller, and move down I bet it will let it go down

**Comment by thadius856** [comment]
You can't scroll because I purposely disabled it. There's times when, if the content gets within 20px of the screen edge, Chrome/Chromium will generate a scroll bar... even though there's nothing to scroll to. This avoids wasting those pixels. You can see it in the CSS at:

```
body {
        ...snip...
    overflow: hidden;
```

Besides, it's a wall-mounted kiosk without a keyboard or touchscreen. You're not meant to scroll it. :)

**Comment by thadius856** [comment]
> And this next one i was able to mock a white border by putting a slightly larger white one right behind it,

I certainly don't like it as much as the previous icon, but we really need a way to dynamically generate. If it's the only way, I'm on board.

**Comment by thadius856** [comment]
Sorry I wasn't on tonight. Was upgrading my office.

Before: my desk - 5 24" IPS panels, merged into one. Not pictured: wife's desk with a single 24" IPS panel, father's desk with two crappy 22" TN monitors.

![pmz9uth](https://f.cloud.github.com/assets/6320180/1928214/cb278588-7e78-11e3-8929-fc650b235a1d.jpg)

After: all desks have 2 24" IPS panels, plus a center 30" IPS panel. This is the only 30" gaming IPS on the market, the U3014 (rev A03!).

![bqcjnzj](https://f.cloud.github.com/assets/6320180/1928219/078704c2-7e79-11e3-85a4-d228c1787e24.jpg)

For kicks 'n' grins, RaspberryPints at 2560x1600:

![untitled](https://f.cloud.github.com/assets/6320180/1928220/17a7e04c-7e79-11e3-99a6-54e3cb46aa3b.png)

Just a "bit" of stretching. ;)

**Comment by JayUnt** [comment]
If you are not a fan, we can still work with the small/large images.  How many different srm images do you think we will need?  We could really keep that one the same way, since we dont have a calculation to figure out the color.

**Comment by thadius856** [comment]
I was hoping for control from 0-40 SRM in .1 steps. Ain't gonna happen of
its manual.

Suppose 1.0 steps is fine.

**Comment by JayUnt** [comment]
Ill look to see if there are any other ways to do it. The problem is the curves of the cup. This is a different beast then the other images that we created, which are all static images, we just show different percentages of it.  Ill let you know if i find a better way.

**Comment by thadius856** [comment]
How how about the outline, plus THREE overlays instead of the one we've been doing: red, green, and blue. Instead of tweaking height, we tweak opacity to mix our own RGB color mix?

**Comment by JayUnt** [comment]
hmm thats a interesting idea.  ill play around with it and let you know what i come up with.

**Comment by Skemp45** [comment]
Thoughts?

http://zookeeperbrewing.com/admin/slider.html

**Comment by Boombrewer** [comment]
Fancy!

Now you guys are just showing off...

But I do like it.

**Comment by JayUnt** [comment]
This is the quick test i did,  I found a brown and found a online converter to change the hex to a rgb percentage.  As you can see in the image, it worked pretty well.  I did the same with a gold, and not so much.  It made it green, which now that i think about it makes sense. since the green they are on top of each other, where rgb are really on the same level.  but with css they can't be on the same level, even if their z-index is the same, they are still layered.

![2014-01-16 11_44_43-srm-test html](https://f.cloud.github.com/assets/35524/1932551/9cf13ba0-7ecd-11e3-8c11-6d1481ac8454.png)

I thought making mask was going to be a pain, so i held off on trying it.  I just made one and saw how simple it is and found i can use the srm image we already have.  here is a test of using the pint glass, the only difference between the two pints is the background color on the container.  I think we might have a winner.

Now we just need to map srm numbers to hex or rgb codes.

![2014-01-16 12_04_08-program manager](https://f.cloud.github.com/assets/35524/1932766/4eca9df6-7ed0-11e3-8df1-9de767ef6e20.png)

**Comment by JayUnt** [comment]
looks like there really is no way to calculate it, but we can import this xml into a table and just use that.

http://brewboard.com/index.php?showtopic=86908&mode=threaded&pid=1035377

**Comment by JayUnt** [comment]
boom! i got it working and commited it.  it uses just 2 image, a mask image and an image for the border (which has to be remade to be more crisp).  The color of the srm is pulled from a new database table srmRgb.

This table has srm to be able to use decimal, i.e. 38.5, but our beers table has arm to be a full number.  Do you want me to change the srm to be a full number or do we want to change beers srm to be a decimal?

The other thing we are going to have to figure out is how to allow them to choose a red if they have fruit punch, or white if they have seltzer.  maybe we can give them another dropdown thats a color override.  By default it will have "use srm" selected, but also have "red", "yellow", "green", "white", etc.  the basic colors.

**Comment by Skemp45** [comment]
I say update the beers srm to allow decimal why not be more accurate???

**Comment by JayUnt** [comment]
my thoughts exactly.  done and done.  We will have to update our's manually, unless you just want to drop everything and re-create.

heres the update query:
ALTER TABLE `raspberrypints`.`beers` CHANGE COLUMN `srm` `srm` DECIMAL(3,1) NOT NULL  ;

**Comment by Skemp45** [comment]
Ran it and it worked great
nice work.

**Comment by Skemp45** [comment]
Im working on converting to pdo 
$conn = new PDO("mysql:host=$dbhost;dbname=$dbname",$dbuser,$dbpass);

**Comment by Skemp45** [comment]
I like the ability to use prepare

**Comment by Skemp45** [comment]
jay 

have you comitted everything for the srm images to work? after altering the db table mine dont show up. 
![errre](https://f.cloud.github.com/assets/6318869/1938204/5cd2260a-7f41-11e3-8c99-120728277529.png)

**Comment by JayUnt** [comment]
That's odd that the first one does but the rest don't, since its just a loop and area l'll using the same images just different background colors. ill take a look when i get to work.

**Comment by thadius856** [comment]
Works fine for me.

**Comment by JayUnt** [comment]
yeah everything is committed on my end.  try doing a shift-refresh to clear your browser cache.

**Comment by thadius856** [comment]
Going to consider this list exhaustive.

New milestones created -- one per release (initial, second, future).

Added all of the to-do list items to each milestone.

Closing this issue. Will be creating new issues for each to-do list item momentarily (for milestone/issue tracking... you'll see!). If they're closed (most will be), you can safely ignore the email spam that is about to ensue.
